WT-6432 Add test case to misuse the timestamp API.

ddanderson opened the pull request.

ddanderson commented:
@sueloverso, I've assigned you for review.  The only tricky part (I think!) is the use of a contextmanager in Python.  Hopefully the comments cover what you need to know, complain if not!

ddanderson commented:
@sueloverso, hold off for the moment.  I discovered that I haven't put in testing for setting global durable_timestamp.  You can still take a look if you want to comment on the direction I'm heading with this test.

sueloverso commented:
I was looking at it now. While you're extending it, my comment was going to be that it would be good if it could include the `timestamp_transaction` API and use that to change the timestamp during a transaction.

sueloverso commented on test/suite/test_timestamp22.py:
```diff
@@ -0,0 +1,298 @@
+
+        # We do a bunch of iterations, doing transactions, prepare, and global timestamp calls
+        # with timestamps that are sometimes valid, sometimes not. We use the iteration number
+        # as a "approximate timestamp", and generate timestamps for our calls that are near
```
'as an'

sueloverso commented on test/suite/test_timestamp22.py:
```diff
@@ -0,0 +1,298 @@
+        # as a "approximate timestamp", and generate timestamps for our calls that are near
+        # that number (within 10).  Thus, as the test runs, the timestamps generally get larger.
+        # We always know the state of global timestamps, so we can predict the success/failure
+        # or each call.
```
'or' should be 'on'

sueloverso commented on test/suite/test_timestamp22.py:
```diff
@@ -0,0 +1,298 @@
+                        session.prepare_transaction(prepare_config)
+
+                with self.expect(self.SUCCESS, 'begin_transaction'):
+                    session.begin_transaction()
```
My earlier comment still stands - we could/should do some checking with `session.timestamp_transaction` instead of or in addition to setting it on the commit call. Is there anything we can do to add testing to changing the timestamp over the course of the transaction with that call? It could be follow-on work to this ticket (if so, perhaps add a comment about adding it, and where/when).

sueloverso reviewed (approved):
Overall LGTM. Two minor typos. One comment about adding more functionality or opening a ticket for it, depending on what you want to do.

ddanderson commented:
@sueloverso, can you please review this again?  I think I've addressed your comments.  Adding timestamp_transaction took longer than I thought, I found an issue that I ticketed.

sueloverso commented on test/suite/test_timestamp22.py:
```diff
@@ -0,0 +1,444 @@
+            # only the durable part fails.  The evidence is that the commit timestamp is set,
+            # as we get a complaint to that effect at the prepare call.  The issue is described
+            # in WT-6995.  This seems wrong, and it's hard to work around the problem, so we'll just
+            # avoid testing that situation for now.  Hence the check for a blank configuration.
```
Minor: I think the comment should have an explicit statement about what change to make once WT-6995 is fixed such as "Once WT-6995 is fixed, remove the entire clause `and config == ''`. And perhaps add a sentence to that ticket reminding the person working on it that this part specifically needs fixing (if it isn't already).

sueloverso commented on test/suite/test_timestamp22.py:
```diff
@@ -0,0 +1,444 @@
+        prepare_config = 'prepare_timestamp=' + timestamp_str(prepare_ts)
+        begin_config = '' if read_ts < 0 else 'read_timestamp=' + timestamp_str(read_ts)
+
+        # We might do a timestamp_transaction calls either before/after inserting
```
Should be `a timestamp_transaction call` (singular) or `might do timestamp_transaction calls`.

sueloverso commented on test/suite/test_timestamp22.py:
```diff
@@ -0,0 +1,444 @@
+            if first_commit_ts < 0:
+                first_commit_ts = running_commit_ts
+
+        # ODDITY: If any setting of the timestamp fail, then an ASSERT will be hit in prepare.
```
`fail` should be `fails`.

sueloverso commented on test/suite/test_timestamp22.py:
```diff
@@ -0,0 +1,444 @@
+
+        # ODDITY: If any setting of the timestamp fail, then an ASSERT will be hit in prepare.
+        # Avoid this, it will crash the test suite when diagnostic mode is enabled, and without
+        # diagnostic mode, the prepare appears to succeed!  (Comment out the if to see it in action).
```
The `if` is commented out. Also there are multiple `if` statements. It isn't clear which one you're referring to or if it supposed to be commented out.

sueloverso commented on test/suite/test_timestamp22.py:
```diff
@@ -0,0 +1,444 @@
+        # ODDITY: If any setting of the timestamp fail, then an ASSERT will be hit in prepare.
+        # Avoid this, it will crash the test suite when diagnostic mode is enabled, and without
+        # diagnostic mode, the prepare appears to succeed!  (Comment out the if to see it in action).
+        # This is reported in WT-7011.  We should fix this to not assert in prepare, and either
```
Although WT-7011 mentions this test, consider if there should be a clear directive that says clearly what should be modified in this test as part of that ticket.

sueloverso commented on test/suite/test_timestamp22.py:
```diff
@@ -0,0 +1,444 @@
+                        self.commit_value = value
+                if needs_rollback:
+                    self.report('rollback_transaction')
+                    needs_rollback = False
```
The multiple uses of `if needs_rollback:` so close together deserve a comment either here or later on. It is a complex depth of potential code paths.

ddanderson replied:
Thanks for catching this.  Turns out I don't need the extra call to do rollback outside of the loop.

sueloverso reviewed (approved):
LGTM. I have a few minor things. Merge when ready.
